[#1272] Implement new Command & Control handling

calohmn opened the pull request:
This is the main implementation for #1272.

### Key classes:
- `ProtocolAdapterCommandConsumerFactoryImpl` (formerly `CommandConsumerFactoryImpl`) is the main implementation class and takes care of the 2 different consumer links (tenant-scoped links and the single link with the adapter instance id).
- `MappingAndDelegatingCommandHandler` is the point where incoming commands, sent on the tenant-scoped address, get handled first. If necessary, commands get forwarded on the adapter instance address.
- `AdapterInstanceCommandHandler` is used for handling commands, having arrived at the adapter instance address.

### Determining mapped gateway and target protocol adapter instance:
This is done in the `CommandTargetMapper` class (formerly `GatewayMapper`).

### Handling of `RegistrationClientFactory`, `DeviceConnectionClientFactory` instances:
The `ProtocolAdapterCommandConsumerFactoryImpl` and `CommandTargetMapperImpl` need such instances. Previously, these ClientFactories got managed in these classes (`connect` method got called).
However, with the new need of `ProtocolAdapterCommandConsumerFactoryImpl` using a `DeviceConnectionClientFactory`, this got somewhat messy, having to override the `ConnectionLifecyle` methods and deciding which of `ProtocolAdapterCommandConsumerFactoryImpl` and `CommandTargetMapperImpl` should call the `connect` method.
Therefore I've changed this so that the ClientFactory instances get passed on in new `initialize` methods.
The management of the ClientFactory instances (calling the `connect` method) gets done in the central `AbstractProtocolAdapterBase` class.

### A change in behaviour:
The `DeviceConnectionService.setCommandHandlingAdapterInstance` method overwrites an existing entry, not returning a "already in use" exception. With this, stale entries cannot block new command subscriptions.
In the same way, `ProtocolAdapterCommandConsumerFactory.createCommandConsumer` doesn't return a `ResourceConflictException` exception (consumer already in use) any more.
That `ResourceConflictException` was inconsistent anyway - command consumer information was not synced across protocol adapter instances.
In the HTTP adapter logic, there was special treatment of that `ResourceConflictException` case. It meant, that with 2 ttd-requests that overlapped each other, the *first* request returned the command message while the second returned nothing.
Now, this is the other way round: The second request will overwrite the command handler and therefore the second request will return the command.
I think this is not a problem, we will just have to document the change in behaviour.

calohmn commented:
Something that I will add still are more unit tests for the new classes.

calohmn commented on client/src/main/java/org/eclipse/hono/client/impl/AdapterInstanceCommandHandler.java:
```diff
@@ -0,0 +1,174 @@
+        if (commandHandler != null) {
+            LOG.trace("using [{}] for received command [{}]", commandHandler, command);
+            // command.isValid() check not done here - it is to be done in the command handler
+            commandHandler.handleCommand(CommandContext.from(command, delivery, currentSpan));
```
The `currentSpan` inside the `CommandContext` is finished as part of the disposition update methods (e.g. `CommandContext.accept()`).

sophokles73 replied:
IMHO the JavaDoc for the *addDeviceSpecificCommandHandler* method should be more explicit about the expectations regarding the passed in handler. In particular, it looks like the handler *must* invoke one of the `CommandContext`'s terminal methods in order to settle the transfer command message transfer and to finish the span, right?

calohmn replied:
Yes. I've adapted the javadoc accordingly, also in the `ProtocolAdapterCommandConsumerFactory`.

calohmn commented on client/src/main/java/org/eclipse/hono/client/impl/CommandTargetMapperImpl.java:
```diff
@@ -0,0 +1,160 @@
+                                }
+                                span.finish();
+                                return Future.succeededFuture(targetInstanceObject);
+                            });
```
Good point.

> we could also move it up one level in the compose hierarchy FMPOV ...

That won't work because of the `viaGateways` list initialized in one compose block.

calohmn replied:
I've put it in a separate method and improved logging/tracing.

calohmn commented on client/src/main/java/org/eclipse/hono/client/impl/CommandTargetMapperImpl.java:
```diff
@@ -0,0 +1,160 @@
+     * @param instancesArray The JSON array containing the target gateway and adapter instance entries to choose from.
+     * @return The chosen JSON object.
+     */
+    protected JsonObject chooseTargetGatewayAndAdapterInstance(final JsonArray instancesArray) {
```
For such a mechanism, the `CommandTargetMapper.getTargetGatewayAndAdapterInstance` method would have to be changed so that it can return a result JSON array.

The implementation of that mechanism would then be where that method is called, ie. in `MappingAndDelegatingCommandHandler.mapAndDelegateIncomingCommand`.

I had decided against such a mechanism mainly because I think it's unclear how a gateway could either say "I as gateway reject the command because I can't handle it" (ie. "try next gateway") or "I reject the command because the target device has rejected it" (ie. "no use in trying other gateways").
One could argue that this is not important, that it wouldn't matter to try the other gateways. But I don't really like that approach, causing traffic and a delay in acknowledging the command for nothing in many cases.

I had rather thought of making it possible to override the `chooseTargetGatewayAndAdapterInstance` method and let a specific implementation better choose the one gateway that the command will be sent to, e.g. by taking extra properties returned in the device connection result into account or by querying the registration service for some gateway device property that helps in deciding which gateway to choose.

sophokles73 replied:
Ok, let's see how we fare with this approach ...

calohmn commented on client/src/main/java/org/eclipse/hono/client/ProtocolAdapterCommandConsumerFactory.java:
```diff
@@ -0,0 +1,126 @@
+     * @param deviceConnectionClientFactory The factory to create a device connection client instance. Note that no
+     *            initialization of this factory will be done here, that is supposed to be done by the calling method.
+     */
+    void initialize(CommandTargetMapper commandTargetMapper,
```
That's how it was before. Like I described in the ticket text above, taking care of the ConnectionLifecycle of the (contained) client factories inside the CommandConsumerFactory makes things complicated.
And taking care of the ConnectionLifecycle is needed as we have the client factory bean methods (in `AbstractAdapterConfig`) annotated with `@Scope("prototype")`, hence dealing with different client factory instances in the places they are used in the `AbstractAdapterConfig`. And as I understood, we want to keep the `@Scope("prototype")` annotation there for the case that the adapter is configured with `maxInstances > 1`, to have each `AbstractProtocolAdapterBase` verticle instance with its own set of client factory and more importantly connection object instances.

sophokles73 replied:
Ok, so the biggest change seems to be that the CommandConsumerFactory now uses the protocol adapter's connections to the Device Registration and Device Connection services instead of using (and managing) its own ones, right?

calohmn replied:
yes

calohmn commented on client/src/main/java/org/eclipse/hono/client/ProtocolAdapterCommandConsumerFactory.java:
```diff
@@ -0,0 +1,126 @@
+     * @return The factory.
+     * @throws NullPointerException if connection or gatewayMapper is {@code null}.
+     */
+    static ProtocolAdapterCommandConsumerFactory create(final HonoConnection connection, final String adapterInstanceId) {
```
You mean the name set here? https://github.com/eclipse/hono/blob/b2096cc0868692a40791e1273b3e44547325518f/core/src/main/java/org/eclipse/hono/connection/impl/ConnectionFactoryImpl.java#L205-L206

Like it is currently implemented, this id changes with each reconnect.

sophokles73 replied:
but that could easily be changed, couldn't it?

calohmn replied:
sure, I'll have a look.

calohmn replied:
I've just committed the corresponding change.

calohmn commented on client/src/main/java/org/eclipse/hono/client/impl/AdapterInstanceCommandHandler.java:
```diff
@@ -0,0 +1,174 @@
+     */
+    public CommandHandlerWrapper getDeviceSpecificCommandHandler(final String tenantId, final String originalDeviceId, final String gatewayId) {
+        // look for a handler with the original device id first
+        CommandHandlerWrapper handler = commandHandlers.get(getDeviceKey(tenantId, originalDeviceId));
```
added null check and javadoc hint.

calohmn commented on client/src/main/java/org/eclipse/hono/client/impl/AdapterInstanceCommandHandler.java:
```diff
@@ -0,0 +1,174 @@
+        final String key = getDeviceKey(tenantId, deviceId);
+        if (commandHandlers.containsKey(key)) {
+            LOG.debug("replacing existing command consumer [tenant-id: {}, device-id: {}]", tenantId, deviceId);
+            // TODO find provide a notification here so that potential resources associated with the replaced consumer can be freed
```
I've changed it to return the old value (a boolean would make me think this is about whether adding the handler succeeded or not). I've also adapted the method name (put..).

calohmn replied:
I've moved the TODO comment here:
https://github.com/eclipse/hono/blob/22f0073941892c831a32b49381de67e038ae5697/client/src/main/java/org/eclipse/hono/client/impl/ProtocolAdapterCommandConsumerFactoryImpl.java#L156
Can be fixed in a subsequent PR, IMHO.

calohmn commented on client/src/main/java/org/eclipse/hono/client/ProtocolAdapterCommandConsumerFactory.java:
```diff
@@ -0,0 +1,126 @@
+    /**
+     * Creates a command consumer for a device.
+     * <p>
+     * For each device only one command consumer may be active at any given time.
```
Adapted the javadoc accordingly.

sophokles73 commented on client/src/main/java/org/eclipse/hono/client/impl/MappingAndDelegatingCommandHandler.java:
```diff
@@ -0,0 +1,288 @@
+/*******************************************************************************
+ * Copyright (c) 2019 Contributors to the Eclipse Foundation
```
2019, 2020?

sophokles73 commented on client/src/main/java/org/eclipse/hono/client/impl/ProtocolAdapterCommandConsumerFactoryImpl.java:
```diff
@@ -0,0 +1,332 @@
+     * @param adapterInstanceId The id of the protocol adapter instance that this factory is running in.
+     * @throws NullPointerException if any of the parameters is {@code null}.
+     */
+    public ProtocolAdapterCommandConsumerFactoryImpl(final HonoConnection connection, final String adapterInstanceId) {
```
I had hoped that we would be able to get rid of the second parameter altogether. But that might be done in a future PR ...

ctron commented on client/src/main/java/org/eclipse/hono/client/CommandTargetMapper.java:
```diff
@@ -0,0 +1,80 @@
+import io.vertx.core.json.JsonObject;
+
+/**
+ * A component for mapping an incoming command to the gateway (if applicable)
```
I think "if applicable" needs a bit more explanation here (and in the following javacdocs of this interface). It think it is an integral aspect, and it would help, explaining a bit more when it might be applicable.

calohmn replied:
I've extended the javadoc here.

sophokles73 replied:
did you?

calohmn replied:
Sorry, I mistook this to be aimed at the javadoc of the `getTargetGatewayAndAdapterInstance` method. I'll update both javadocs (class and method)...

calohmn replied:
done

calohmn commented:
(Rebased. I'm about to commit more tests and a small fix still.)

sophokles73 commented on client/src/main/java/org/eclipse/hono/client/impl/MappingAndDelegatingCommandHandler.java:
```diff
@@ -0,0 +1,288 @@
+        if (!originalCommand.isValid()) {
+            originalCommandContext.reject(new ErrorCondition(Constants.AMQP_BAD_REQUEST, "malformed command message"));
+            return;
+        }
```
it doesn't look like this can happen, can it?

sophokles73 commented on client/src/main/java/org/eclipse/hono/client/impl/MappingAndDelegatingCommandHandler.java:
```diff
@@ -0,0 +1,288 @@
+            if (commandTargetResult.succeeded()) {
+                final String targetDeviceId = commandTargetResult.result().getString(DeviceConnectionConstants.FIELD_PAYLOAD_DEVICE_ID);
+                final String targetAdapterInstance = commandTargetResult.result().getString(DeviceConnectionConstants.FIELD_ADAPTER_INSTANCE_ID);
+
```
this is the same code as in the main method, isn't it? would it be possible to change the behavior so that we first try to delegate processing of a command to a local handler in any case (regardless of whether a command is valid or not) and then, if valid, forward it to the correct adapter instance?

calohmn replied:
Good point. I've refactored the method here and made that change.

calohmn commented on client/src/main/java/org/eclipse/hono/client/impl/CommandHandlerWrapper.java:
```diff
@@ -24,28 +24,36 @@
-            final Handler<CommandContext> commandHandler, final Handler<Void> remoteCloseHandler) {
+    public CommandHandlerWrapper(final String tenantId, final String deviceId, final String gatewayId,
+                                 final Handler<CommandContext> commandHandler) {
+        this.tenantId = Objects.requireNonNull(tenantId);
```
done

calohmn commented on client/src/main/java/org/eclipse/hono/client/impl/DelegatedCommandSenderImpl.java:
```diff
@@ -225,21 +222,19 @@ private static Message createDelegatedCommandMessage(final Message originalMessa
         Objects.requireNonNull(con);
 
-        final String targetAddress = getTargetAddress(tenantId, deviceId);
+        final String targetAddress = getTargetAddress(adapterInstanceId);
```
Added null check and javadoc info.

calohmn commented:
Added tests and removed a minor inconsistency regarding the `AdapterInstanceCommandHandler.getDeviceSpecificCommandHandler` method (now there's no `gatewayId` parameter in there anymore).

sophokles73 commented on client/src/main/java/org/eclipse/hono/client/CommandTargetMapper.java:
```diff
@@ -0,0 +1,110 @@
+ * A component for determining where a incoming command message should be targeted at when processing the message in a
+ * protocol adapter.
+ * <p>
+ * This refers to finding the <em>protocol handler instance</em> that a device or gateway has connected to in order to
```
the protocol *adapter* instance?

calohmn replied:
fixed
